File: server/internal/model/dto/node.go
/*
 * @Date: 2025-06-19 00:02:51
 * @LastEditors: peng pgs1108pgs@gmail.com
 * @LastEditTime: 2025-06-22 21:39:39
 * @FilePath: /thinking-map/server/internal/model/dto/node.go
 */
package dto

import (
	"time"

	"github.com/PGshen/thinking-map/server/internal/model"
)

const (
	ConversationTypeDecomposition = "decomposition"
	ConversationTypeConclusion    = "conclusion"
)

// CreateNodeRequest represents the request body for creating a node
type CreateNodeRequest struct {
	MapID    string         `json:"MapID" binding:"required,uuid"`
	ParentID string         `json:"parentID" binding:"required,uuid"`
	NodeType string         `json:"nodeType" binding:"required"`
	Question string         `json:"question" binding:"required,max=500"`
	Target   string         `json:"target" binding:"max=500"`
	Position model.Position `json:"position" binding:"required"`
}

// UpdateNodeRequest represents the request body for updating a node
type UpdateNodeRequest struct {
	Question string         `json:"question" binding:"max=500"`
	Target   string         `json:"target" binding:"max=500"`
	Position model.Position `json:"position"`
}

type UpdateNodeContextRequest struct {
	Context model.DependentContext `json:"context"`
}

// NodeResponse represents the node data in responses
type NodeResponse struct {
	ID            string                 `json:"id"`
	NodeID        string                 `json:"nodeID"`
	MapID         string                 `json:"mapID,omitempty"`
	ParentID      string                 `json:"parentID"`
	NodeType      string                 `json:"nodeType"`
	Question      string                 `json:"question"`
	Target        string                 `json:"target"`
	Context       model.DependentContext `json:"context"`
	Decomposition model.Decomposition    `json:"decomposition"`
	Conclusion    model.Conclusion       `json:"conclusion"`
	Status        string                 `json:"status"`
	Position      model.Position         `json:"position"`
	Metadata      interface{}            `json:"metadata"`
	Dependencies  model.Dependencies     `json:"dependencies"`
	CreatedAt     time.Time              `json:"createdAt"`
	UpdatedAt     time.Time              `json:"updatedAt"`
}

// NodeListResponse represents the list of nodes in a map
type NodeListResponse struct {
	Nodes []NodeResponse `json:"nodes"`
}

// modelToNodeResponse 将model.ThinkingNode转为dto.NodeResponse
func ToNodeResponse(n *model.ThinkingNode) NodeResponse {
	return NodeResponse{
		ID:            n.ID,
		MapID:         n.MapID,
		ParentID:      n.ParentID,
		NodeType:      n.NodeType,
		Question:      n.Question,
		Target:        n.Target,
		Context:       n.Context,
		Decomposition: n.Decomposition,
		Conclusion:    n.Conclusion,
		Status:        n.Status,
		Position:      n.Position,
		Metadata:      n.Metadata,
		Dependencies:  n.Dependencies,
		CreatedAt:     n.CreatedAt,
		UpdatedAt:     n.UpdatedAt,
	}
}

File: web/src/types/node.ts
/*
 * @Date: 2025-07-07 22:05:27
 * @LastEditors: peng pgs1108pgs@gmail.com
 * @LastEditTime: 2025-07-08 00:39:50
 * @FilePath: /thinking-map/web/src/types/node.ts
 */
// 与后端 dto/node.go 对齐的节点类型定义
import { MessageResponse } from './message';
import type { ApiResponse } from './response';

export interface Position {
  x: number;
  y: number;
}

export interface Node {
  id: string;
  label: string;
  // 可根据实际需求扩展属性
}

export interface Edge {
  id: string;
  source: string;
  target: string;
  // 可根据实际需求扩展属性
}

export interface CustomNodeModel {
  id: string;
  parentID?: string;
  nodeType: string;
  question: string;
  target: string;
  status: 'initial' | 'pending' | 'running' | 'completed' | 'error';
  context?: DependentContext;
  decomposition?: Decomposition;
  conclusion?: Conclusion;
  metadata?: any;
  selected?: boolean;
  isEditing?: boolean;
  // 交互事件（由外部注入，非持久数据）
  onEdit?: (mapID: string|null, id: string, data: Partial<CustomNodeModel>) => Promise<NodeResponse>;
  onDelete?: (mapID: string|null, id: string) => void;
  onAddChild?: (id: string) => void;
  onSelect?: (id: string) => void;
  onDoubleClick?: (id: string) => void;
  onContextMenu?: (id: string, e: React.MouseEvent) => void;
  onUpdateID?: (mapID: string|null, oldID: string, newID: string) => void;
}

export interface DependentContext {
  ancestor: NodeContextItem[];
  prevSibling: NodeContextItem[];
  children: NodeContextItem[];
}

// 节点上下文项类型
export interface NodeContextItem {
  question: string;
  target: string;
  conclusion?: string;
  abstract?: string;
  status: string;
}

export interface Decomposition {
  isDecompose: boolean;
  lastMessageID: string;
  messages: MessageResponse[];
}

export interface Conclusion {
  lastMessageID: string;
  content: string;
  messages: MessageResponse[];
}

export interface NodeResponse {
  id: string;
  mapID: string;
  parentID: string;
  nodeType: string;
  question: string;
  target: string;
  context: {
    ancestor: NodeContextItem[];
    prevSibling: NodeContextItem[];
    children: NodeContextItem[];
  };
  decomposition: Decomposition;
  conclusion: Conclusion;
  status: string;
  position: Position;
  metadata: Record<string, any>;
  createdAt: string;
  updatedAt: string;
}

// 创建节点请求参数
export interface CreateNodeRequest {
  mapID: string;
  parentID: string;
  nodeType: string;
  question: string;
  target: string;
  position: Position;
}

// 更新节点请求参数
export interface UpdateNodeRequest {
  question?: string;
  target?: string;
  position?: Position;
}

// 更新节点上下文请求参数
export interface UpdateNodeContextRequest {
  context: {
    ancestor: NodeContextItem[];
    prevSibling: NodeContextItem[];
    children: NodeContextItem[];
  };
}

export type NodeListResponse = ApiResponse<{
  nodes: NodeResponse[];
}>;
export type CreateNodeResponse = ApiResponse<NodeResponse>;
export type UpdateNodeResponse = ApiResponse<NodeResponse>;
export type DeleteNodeResponse = ApiResponse<null>;
export type UpdateNodeContextResponse = ApiResponse<NodeResponse>;
export type ResetNodeContextResponse = ApiResponse<NodeResponse>;